### PR TITLE
feat(search): update search pattern matching rules

### PR DESCRIPTION
**Pull request by laurentiuNiculae**
**What type of PR is this?**

**Which issue does this PR fix**:
#909 

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/meta/repodb/common/common.go [hunk 1]
```diff
@@ -71,34 +77,47 @@ func ScoreRepoName(searchText string, repoName string) int {
...
 	for i := 0; i < len(searchTextSlice)-1; i++ {
 		if searchTextSlice[i] != repoNameSlice[i] {
-			return -1
+			begginingMatches = false
+
+			break
+		}
+	}
+
+	if begginingMatches {
```
Comment (rchincha) [inline pkg/meta/repodb/common/common.go:110]: typo

beginning

File: pkg/meta/repodb/common/common.go [hunk 2]
```diff
@@ -62,6 +62,12 @@ func ValidateRepoReferenceInput(repo, reference string, manifestDigest godigest.
 	return nil
 }
 
+const (
```
Comment (rchincha) [inline pkg/meta/repodb/common/common.go:65]: what is this for?

File: pkg/meta/repodb/common/common.go [hunk 3]
```diff
@@ -71,34 +77,47 @@ func ScoreRepoName(searchText string, repoName string) int {
 	}
```
Comment (andaaron) [inline pkg/meta/repodb/common/common.go:77]: Make sure and exact match always has lowest score.

File: pkg/meta/repodb/common/common.go [hunk 4]
```diff
@@ -71,34 +77,47 @@ func ScoreRepoName(searchText string, repoName string) int {
...
 
 	if len(searchTextSlice) == 1 {
-		// check if it maches first or last name in path
+		// check if it maches first or last directory name in path, we'll give priority for these names
```
Comment (andaaron) [inline pkg/meta/repodb/common/common.go:80]:
```suggestion
		// check if it matches last directory name in path, we'll give priority for these names
```

**Review by andaaron — APPROVED** [review]

File: pkg/meta/repodb/boltdb-wrapper/boltdb_wrapper.go [hunk 5]
```diff
@@ -999,7 +999,7 @@ func (bdw *DBWrapper) SearchRepos(ctx context.Context, searchText string, filter
 				return err
 			}
 
-			if score := common.ScoreRepoName(searchText, string(repoName)); score != -1 {
+			if score := common.RankRepoName(searchText, string(repoName)); score != -1 {
```
Comment (rchincha) [inline pkg/meta/repodb/boltdb-wrapper/boltdb_wrapper.go:1002]: call "score" rank also?

**Comment by rchincha** [comment]
So what is the overall logic for ranking? image:tag first, repo names, and any wildcards? We should eventually also allow searching within annotations. This use case is coming up often.

**Comment by andaaron** [comment]
> So what is the overall logic for ranking? image:tag first, repo names, and any wildcards? We should eventually also allow searching within annotations. This use case is coming up often.

- If a tag is specified (`:`), the search string for the repo needs to be an exact match. That is a tag search, not a repo search. The logic for that was not changed in this PR.
- If a `:` is not specified, the repo string is compared to the search string according to the new comments.

We did not decide on any wildcard, or annotation searches to my knowledge.
At the moment the presence of `:` signals if this is a repo or a tag search.

What would be a good example of a string the user sends as an annotation search? How would the servers know that is an annotation search, not a repo or a tag search?

Actually I am not convinced searches by annotations are a good idea: which annotations?, on the manifest itself?, on an index?, on descriptors inside a manifest or index?, by annotation key? by annotation value?, by an annotation value on a specific key?

Why would I want to search on an annotation anyway? What is the use case?

**Review by andaaron — DISMISSED** [review]
It's good to see we removed `score` from the graphql api, as we had no benefit in exposing it.

File: pkg/meta/repodb/common/common.go [hunk 6]
```diff
@@ -62,43 +62,89 @@ func ValidateRepoReferenceInput(repo, reference string, manifestDigest godigest.
...
+
+// RankRepoName associates a rank to a given repoName given a searchText.
+// The imporance of the value grows inversly proportional to the int value it has.
+// For example: rank(1) > rank(10) > rank(100)...
+func RankRepoName(searchText string, repoName string) int {
+	searchText = strings.Trim(searchText, "/")
 	searchTextSlice := strings.Split(searchText, "/")
 	repoNameSlice := strings.Split(repoName, "/")
 
+	// ex: zot/test/alpine, "alpine"
```
Comment (rchincha) [inline pkg/meta/repodb/common/common.go:83]: "repo/repo1/abc" use other names
or remove the comment, if there are examples in test cases

**Review by andaaron — DISMISSED** [review]

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**